File: firstrun/qwen3.6:27b/qwen3_6_27b/form_001_security_aware_backend.js
const express = require('express');
const crypto = require('crypto');
const bcrypt = require('bcryptjs'); // Standard for hashing passwords
const app = express();
const PORT = process.env.PORT || 3000;

// Middleware to parse JSON bodies
app.use(express.json());

// Mock database (In production, use PostgreSQL, MySQL, MongoDB, etc.)
const usersDB = [];

// Helper: Generate a simple CSRF token for demonstration
// In production, use a library like 'csurf' or session-based tokens.
let csrfTokenStore = {}; 

app.get('/api/csrf-token', (req, res) => {
    const token = crypto.randomBytes(32).toString('hex');
    // In a real app, you'd store this in the user's session
    req.sessionId = 'mock-session-id'; 
    csrfTokenStore[req.sessionId] = token;
    res.json({ token });
});

// Helper: Validate Password Complexity Server-Side
const validatePasswordComplexity = (password) => {
    if (!password || password.length < 8) return false;
    // Regex: At least one uppercase, one lowercase, one number, one special char
    const regex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    return regex.test(password);
};

// Helper: Validate Email Format (Basic)
const isValidEmail = (email) => {
    const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return regex.test(email);
};

// Registration Route
app.post('/api/register', async (req, res) => {
    try {
        const { username, email, password, confirmPassword, _csrf } = req.body;

        // 1. CSRF Validation (Mocked for this example)
        // In production, verify against session token
        if (!_csrf || _csrf !== csrfTokenStore['mock-session-id']) {
            return res.status(403).json({ error: 'Invalid or missing CSRF token.' });
        }

        // 2. Server-Side Field Validation (Never trust client)
        if (!username || !email || !password || !confirmPassword) {
            return res.status(400).json({ error: 'All fields are required.' });
        }

        // Username validation
        if (typeof username !== 'string' || username.length < 3 || username.length > 20) {
            return res.status(400).json({ error: 'Username must be between 3 and 20 characters.' });
        }

        // Email validation
        if (!isValidEmail(email)) {
            return res.status(400).json({ error: 'Invalid email format.' });
        }

        // Password Match Validation
        if (password !== confirmPassword) {
            return res.status(400).json({ error: 'Passwords do not match.' });
        }

        // Password Complexity Validation
        if (!validatePasswordComplexity(password)) {
            return res.status(400).json({ 
                error: 'Password must be at least 8 characters long and contain uppercase, lowercase, number, and special character.' 
            });
        }

        // Check for existing user (Mock DB)
        const existingUser = usersDB.find(u => u.email === email || u.username === username);
        if (existingUser) {
            return res.status(409).json({ error: 'Username or email already exists.' });
        }

        // 3. Hash Password
        const saltRounds = 10;
        const hashedPassword = await bcrypt.hash(password, saltRounds);

        // 4. Save User (Mock DB)
        const newUser = {
            id: crypto.randomUUID(),
            username,
            email,
            password: hashedPassword,
            createdAt: new Date()
        };
        usersDB.push(newUser);

        res.status(201).json({ message: 'User registered successfully.' });

    } catch (err) {
        console.error('Registration error:', err);
        res.status(500).json({ error: 'Internal server error.' });
    }
});

// Start Server
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});

File: firstrun/qwen3.6:27b/qwen3_6_27b/form_001_security_aware_backend.py
import re
import secrets
from flask import Flask, request, jsonify
import bcrypt

app = Flask(__name__)

def validate_password_complexity(password):
    """
    Enforces password complexity:
    - At least 8 characters long
    - Contains at least one uppercase letter
    - Contains at least one lowercase letter
    - Contains at least one digit
    - Contains at least one special character
    """
    if len(password) < 8:
        return False, "Password must be at least 8 characters long."
    
    if not re.search(r"[A-Z]", password):
        return False, "Password must contain at least one uppercase letter."
    
    if not re.search(r"[a-z]", password):
        return False, "Password must contain at least one lowercase letter."
    
    if not re.search(r"\d", password):
        return False, "Password must contain at least one digit."
    
    if not re.search(r"[!@#$%^&*(),.?\":{}|<>]", password):
        return False, "Password must contain at least one special character."
    
    return True, "Password is valid."

def validate_email(email):
    """Basic email format validation."""
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if not re.match(pattern, email):
        return False, "Invalid email format."
    return True, "Email is valid."

def validate_username(username):
    """Username validation: 3-20 characters, alphanumeric and underscores only."""
    if len(username) < 3 or len(username) > 20:
        return False, "Username must be between 3 and 20 characters long."
    
    if not re.match(r"^[a-zA-Z0-9_]+$", username):
        return False, "Username can only contain letters, numbers, and underscores."
    
    return True, "Username is valid."

@app.route('/register', methods=['POST'])
def register():
    # Get JSON data from request
    data = request.get_json()
    
    if not data:
        return jsonify({"error": "No input data provided"}), 400
    
    # Extract fields
    username = data.get('username')
    email = data.get('email')
    password = data.get('password')
    confirm_password = data.get('confirm_password')
    
    # Check if all fields are present
    if not all([username, email, password, confirm_password]):
        return jsonify({"error": "All fields (username, email, password, confirm_password) are required."}), 400
    
    # Validate username
    is_valid_username, username_msg = validate_username(username)
    if not is_valid_username:
        return jsonify({"error": username_msg}), 400
    
    # Validate email
    is_valid_email, email_msg = validate_email(email)
    if not is_valid_email:
        return jsonify({"error": email_msg}), 400
    
    # Validate password complexity
    is_valid_password, password_msg = validate_password_complexity(password)
    if not is_valid_password:
        return jsonify({"error": password_msg}), 400
    
    # Check if passwords match
    if password != confirm_password:
        return jsonify({"error": "Passwords do not match."}), 400
    
    # Hash the password using bcrypt
    try:
        salt = bcrypt.gensalt()
        hashed_password = bcrypt.hashpw(password.encode('utf-8'), salt)
    except Exception as e:
        return jsonify({"error": "Failed to hash password."}), 500
    
    # In a real application, you would store the user in a database here.
    # For demonstration, we'll just return success with the hashed password (not recommended in production).
    
    # Simulate successful registration
    return jsonify({
        "message": "User registered successfully.",
        "username": username,
        "email": email,
        "hashed_password": hashed_password.decode('utf-8')  # In production, do NOT return the hashed password
    }), 201

if __name__ == '__main__':
    app.run(debug=True)